### DISCUSSION TITLE
AutoCloseable or self-disposing temp files

### DISCUSSION THREAD

**Discussion by mpollmeier** [lihaoyi/os-lib#140] [Ideas]
For temp files/dirs we currently rely on the JVM's `deleteOnExit` shutdown handler to remove them. That's not ideal though for a number of reasons, including:
* some systems may run forever, therefor never letting removing any temp files
* shutdown process is slowed down
* files stay around forever when process gets killed

There's a variety of ways we could handle this better - two that come to mind immediately are
1) implement `AutoCloseable`, e.g. on `Path` or add a new type `TempPath` that does it. That way users can rely on things like `scala.util.Using`. 
2) provide our own `using` abstraction, similar to what [better-files does](https://github.com/pathikrit/better-files/blob/88d55db/core/src/main/scala/better/files/File.scala#L1275-L1289) - see https://github.com/pathikrit/better-files#temporary-files

Any thoughts on this?

**Comment by lefou** [comment]
I like the `TempPath` type very much. I would use the existing `AutoClosable`, so that it works with `Using`. 

We could also add some convenience methods `withTempFile` and `withTempDir`, which I typically use in tests (https://github.com/lefou/LambdaTest#testing-files-and-directories-with-tempfile). Whereas in tests, I prefer some more fine control over the lifetime, e.g. when the test fails, I'd like to keep and print the temp file, so I can analyze it further.

**Reply by mpollmeier** [reply]
Sounds good, that'd be my preference as well. I won't be able to use a keyboard for a week or two, so unless anyone else jumps in, I'll send a PR afterwards.

**Comment by mpollmeier** [comment]
here's a first attempt: https://github.com/com-lihaoyi/os-lib/pull/147
